### PR TITLE
#27 refactoring slack parsed command and slack name handler service

### PR DESCRIPTION
**Pull request by ksergey12**
connect to #27

### CODE REVIEW

File: src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java [hunk 1]
```diff
@@ -18,24 +17,26 @@
...
     public void setup() {
-        users = new HashMap<>();
-        users.put("@from", new UserDTO("uuid0", "@from"));
-        users.put("@slack1", new UserDTO("uuid1", "@slack1"));
-        users.put("@slack2", new UserDTO("uuid2", "@slack2"));
+        fromUser = new UserDTO("uuid0", "@from");
+        usersInText = new ArrayList<>();
+        usersInText.add(new UserDTO("uuid1", "@slack1"));
+        usersInText.add(new UserDTO("uuid2", "@slack2"));
+        usersInText.add(new UserDTO("uuid3", "@slack3"));
```
Comment (NikolayNN) [inline src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java]: I think the list for each test should be contains specific users accordance with the text.
for example 
"text text @slack1 text" 
usersInText must contains only one object UserDTO

File: src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java [hunk 2]
```diff
@@ -91,22 +94,19 @@ public void getFromUser() {
     @Test
     public void getUserCount() {
         //given
-        String text1 = "text @slack text";
-        String text2 = "@slack text @slack text @slack";
+        String text = "text @slack text";
```
Comment (NikolayNN) [inline src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java]: the text contains only one slackName, but in the test
assertEquals(3, slackParsedCommand.getUserCountInText());

**Review by NikolayNN — CHANGES_REQUESTED** [review]

File: src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java (hunk 1 above)
Comment (ksergey12) [reply src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java]: Fixed.

File: src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java (hunk 2 above)
Comment (ksergey12) [reply src/test/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommandTest.java]: Fixed.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 3]
```diff
@@ -3,93 +3,63 @@
...
-        }
-        this.fromSlackName = fromSlackName;
+    public SlackParsedCommand(UserDTO fromUser, String text, List<UserDTO> usersInText) {
+        this.fromUser = fromUser;
         this.text = text;
-        this.users = users;
-        slackNamesInText = getSlackNames();
-        userCount = slackNamesInText.size();
+        this.usersInText = usersInText;
+        userCountInText = usersInText.size();
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: delete the field userCountInText

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 4]
```diff
@@ -3,93 +3,63 @@
...
-        }
-        return result;
-    }
-
-    private void checkSlackNamesPresence() {
-        if (userCount == 0) {
-            logger.warn("The text: [{}] doesn't contain slack name", text);
-            throw new WrongCommandFormatException(String.format("The text '%s' doesn't contain any slackName", text));
-        }
+        return userCountInText;
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: return usersInText.size();

**Review by NikolayNN — CHANGES_REQUESTED** [review]

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 3 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 4 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java [hunk 5]
```diff
@@ -7,89 +7,76 @@
...
-        if (userCount == 0) {
-            logger.warn("The text: [{}] doesn't contain slack name", text);
-            throw new WrongCommandFormatException(String.format("The text '%s' doesn't contain any slackName", text));
-        }
+    @Override
+    public int hashCode() {
+        int result = fromUser != null ? fromUser.hashCode() : 0;
+        result = 31 * result + (text != null ? text.hashCode() : 0);
+        result = 31 * result + (usersInText != null ? usersInText.hashCode() : 0);
+        return result;
```
Comment (NikolayNN) [inline src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: You can just use  anotation  @EqualsAndHashCode of lombok.
And add the Annotation to UserDTO

File: src/test/java/ua/com/juja/microservices/keepers/slackbot/service/impl/SlackNameHandlerServiceTest.java [hunk 6]
```diff
@@ -43,48 +45,47 @@ public void setup() {
...
+        SlackParsedCommand expected = new SlackParsedCommand(userFrom, text, Collections.singletonList(user1));
         //when
-        SlackParsedCommand slackParsedCommand = slackNameHandlerService.createSlackParsedCommand(userFrom.getSlack(), text);
+        SlackParsedCommand actual = slackNameHandlerService.createSlackParsedCommand(userFrom.getSlack(), text);
         //then
-        assertEquals("SlackParsedCommand(fromSlackName=@slackFrom, text=text @slack1 TexT text., " +
-                "users={@slackFrom=UserDTO(uuid=AAA000, slack=@slackFrom), @slack1=UserDTO(uuid=AAA111, " +
-                "slack=@slack1)}, slackNamesInText=[@slack1], userCount=1)", slackParsedCommand.toString());
+        assertEquals(expected, actual);
+        assertTrue(expected.hashCode() == actual.hashCode());
```
Comment (NikolayNN) [inline src/test/java/ua/com/juja/microservices/keepers/slackbot/service/impl/SlackNameHandlerServiceTest.java]: it is redaundant check, if object equals then hashCode equals

**Review by NikolayNN — CHANGES_REQUESTED** [review]

File: src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java (hunk 5 above)
Comment (ksergey12) [reply src/main/java/ua/com/juja/microservices/keepers/slackbot/model/SlackParsedCommand.java]: Done.

File: src/test/java/ua/com/juja/microservices/keepers/slackbot/service/impl/SlackNameHandlerServiceTest.java (hunk 6 above)
Comment (ksergey12) [reply src/test/java/ua/com/juja/microservices/keepers/slackbot/service/impl/SlackNameHandlerServiceTest.java]: Fixed.

**Review by NikolayNN — APPROVED** [review]

**Review by OlgaKulikova — APPROVED** [review]

**Pull request merged**